pkg/controller/ephemeraljob/ephemeraljob_controller.go

```diff
@@ -0,0 +1,350 @@
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to EphemeralJob
+	err = c.Watch(&source.Kind{Type: &appsv1alpha1.EphemeralJob{}}, &handler.EnqueueRequestForObject{}, predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			old := e.ObjectOld.(*appsv1alpha1.EphemeralJob)
+			new := e.ObjectNew.(*appsv1alpha1.EphemeralJob)
+			// only update this 4 fields
+			if old.Spec.TTLSecondsAfterCreated != new.Spec.TTLSecondsAfterCreated ||
+				old.Spec.Paused != new.Spec.Paused || old.Spec.Parallelism != new.Spec.Parallelism ||
+				old.Spec.Replicas != new.Spec.Replicas {
+
+				klog.V(3).Infof("Observed updated Spec for ephemeral job: %s/%s", new.Namespace, new.Name)
+				return true
+			}
+			return false
+		},
+	})
+	if err != nil {
+		return err
+	}
+	// Watch for changes to EphemeralJob
+	err = c.Watch(&source.Kind{Type: &v1.Pod{}}, &podHandler{mgr.GetCache()})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileEphemeralJob{}
+
+// ReconcileImagePullJob reconciles a ImagePullJob object
+type ReconcileEphemeralJob struct {
+	client.Client
+	scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=apps.kruise.io,resources=imagepulljobs,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=apps.kruise.io,resources=imagepulljobs/status,verbs=get;update;patch
+
+// Reconcile reads that state of the cluster for a ImagePullJob object and makes changes based on the state read
+// and what is in the ImagePullJob.Spec
+// Automatically generate RBAC rules to allow the Controller to read and write Deployments
+func (r *ReconcileEphemeralJob) Reconcile(request reconcile.Request) (res reconcile.Result, err error) {
+	start := time.Now()
+	klog.V(5).Infof("Starting to process EphemeralJob %v", request.Name)
+	var requeueMsg string
+	defer func() {
+		if err != nil {
+			klog.Warningf("Failed to process EphemeralJob %v err %v, elapsedTime %v", request.Name, time.Since(start), err)
+		} else if res.RequeueAfter > 0 {
+			klog.Infof("Finish to process EphemeralJob %v, elapsedTime %v, RetryAfter %v: %v", request.Name, time.Since(start), res.RequeueAfter, requeueMsg)
+		} else {
+			klog.Infof("Finish to process EphemeralJob %v, elapsedTime %v", request.Name, time.Since(start))
+		}
+	}()
+
+	job := &appsv1alpha1.EphemeralJob{}
+	err = r.Get(context.TODO(), request.NamespacedName, job)
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		klog.Errorf("Failed to get ephemeral job %s,", job.Name)
+		return reconcile.Result{}, err
+	}
+
+	// The Job has been finished
+	if job.Spec.TTLSecondsAfterCreated != nil && job.CreationTimestamp.Add(time.Duration(*job.Spec.TTLSecondsAfterCreated)*time.Second).Before(time.Now()) {
```

Comment:
job.CreationTimestamp.Add(time.Duration(*job.Spec.TTLSecondsAfterCreated)*time.Second).Before(time.Now())
 
-> time.Since(job.CreationTimestamp) > time.Duration(*job.Spec.TTLSecondsAfterCreated)*time.Second